packages/vats/src/lib-board.js

```diff
@@ -60,81 +177,144 @@ function makeBoard(
+  );
+
+  return { readonlyMarshaller, publishingMarshaller };
+};
 
-  const board = Far('Board', {
-    getPublishingMarshaller: () => publishingMarshaller,
-    getReadonlyMarshaller: () => readonlyMarshaller,
-    // Add if not already present
-    getId: value => {
-      if (!valToId.has(value)) {
-        lastSequence += 1n;
-        const seq = lastSequence;
-
-        // Append the CRC, so that the last part of the board ID is
-        // well-distributed.
-        const crcInput = `${prefix}${seq}`;
-        const crc = calcCrc(crcInput, crcDigits);
-
-        const id = `${prefix}${crc}${seq}`;
-
-        valToId.init(value, id);
-        idToVal.init(id, value);
-      }
-      return valToId.get(value);
+/**
+ * A board is a two-way mapping between objects (such as issuers,
+ * brands, installation handles) and plain-data string IDs.
+ * A well-known board allows sharing objects by way of their ids.
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage for upgrade successors
+ * @see {@link packages/SwingSet/docs/vat-upgrade.md|Vat Upgrade}
+ */
+export const prepareBoardClass = baggage => {
+  // ephemeral marshallers, keyed by `this` virtual object identity
+  const my = makeEphemeraProvider(({ state }) => makeMarshallers(state));
+
+  return prepareExoClassKit(baggage, 'Board', BoardI, makeInit(baggage), {
+    board: {
+      /**
+       * Provide an ID for a value, generating one if not already present.
+       *
+       * Each board ID includes a CRC so that the last part is well-distributed,
+       * which mitigates typo risks.
+       *
+       * Scaling Consideration: since we cannot know when consumers
+       * are no longer interested in an ID, a board holds a strong reference
+       * to `value` for its entire lifetime. For a well-known board, this
+       * is essentially forever.
+       *
+       * @param {Key} value
+       * @throws if `value` is not a Key in the @agoric/store sense
+       */
+      getId(value) {
+        return getId(value, this.state);
+      },
+      /**
+       * @param {string} id
+       * @throws if id is not in the mapping
+       */
+      getValue(id) {
+        return getValue(id, this.state);
+      },
+      /**
+       * Adapter for lookup() protocol.
+       *
+       * @param  {...string} path
+       */
+      // XXX which lookup() protocol is that?
+      async lookup(...path) {
+        const {
+          facets: { board },
+        } = this;
+        if (path.length === 0) {
+          return board;
+        }
+        const [first, ...rest] = path;
+        const firstValue = board.getValue(first);
+        if (rest.length === 0) {
+          return firstValue;
+        }
+        return E(firstValue).lookup(...rest);
+      },
+      /** @param {Key} val */
+      has(val) {
+        const { state } = this;
+        return state.valToId.has(val);
+      },
+      ids() {
+        const { state } = this;
+        // XXX O(n) compute for O(1) size message XXX
```

Comment:
most XXX tags are devex or code factoring. this one affects performance and scalability so maybe a PERF or SCALE tag (could be in addition to XXX)

Reply:
IOU an issue